node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,637 @@
+/*
+ * Copyright (c) 2017, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert         = require("chai").assert;
+const co             = require("co");
+const NodeGit        = require("nodegit");
+
+const Commit              = require("./commit");
+const DoWorkQueue         = require("./do_work_queue");
+const GitUtil             = require("./git_util");
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const SubmoduleUtil       = require("./submodule_util");
+const TreeUtil            = require("./tree_util");
+
+const FILEMODE            = NodeGit.TreeEntry.FILEMODE;
+
+/**
+ * Return a string having the value of the specified `sha` with the "/"
+ * character inserted between the second and third characters of `sha`.  The
+ * behavior is undefined unless `sha` is at least three characters long.
+ *
+ * @param {String} sha
+ * @return {String}
+ */
+exports.splitSha = function (sha) {
+    const pre = sha.slice(0, 2);
+    const post = sha.slice(2);
+    return `${pre}/${post}`;
+};
+
+/**
+ * Return the name of the ref used to indicate that the commit having the
+ * specified `sha` has beren converted.
+ *
+ * @param {String} sha
+ * @return {String}
+ */
+exports.convertedRefName = function (sha) {
+    return "refs/stitched/converted/" + exports.splitSha(sha);
+};
+
+/**
+ * Return the name of the ref indicating that the submodule commit having the
```

Comment:
Wait, why do we need the submodule commit to be under the meta sha?

Reply:
It's mapping from meta-repo commit sha to stitched commit sha

Reply:
No it isn't.

Reply:
OK, I (originally) thought you were talking about `convertedRefName`, but you're referring to next function, `fetchedSubRefName`. What I'm doing here is making a ref for submodule commits I fetch so that they don't get GC'd. I'll make this more clear in the comments.

Reply:
Couldn't we just use an ordinary synthetic ref?

Reply:
Yes, that makes sense.